SR-648: Allow swiftpm to statically link binaries on Linux

spevans opened the pull request:
This PR requires https://github.com/apple/swift-corelibs-foundation/pull/1011 and 
https://github.com/apple/swift-corelibs-xctest/pull/190 to be merged beforehand.

- If --build-swift-static-stdlib option is used then also produce
  static versions of libXCTest.a and libdispatch.a and put them
  into the lib/swift_static/linux/ toolchain directory.
  libFoundation.a is already being built and deployed there.

- Binaries with the swift libs statically linked in can then be
  built with using the command:

  swift build -Xswiftc -static-stdlib -Xlinker -lcurl -Xlinker -lxml2 -Xlinker -lbsd -Xlinker --allow-multiple-definition

  Note: This is a dynamic binary with the libswiftCore,  libFoundation and libdispatch libraries statically linked in.

- Further fixes should reduce the complexity of the above command.

There arent any unit tests in this PR, however I will add a followup one which builds a static version of TestFoundation to test xctest, libdispatch and libfoundation together.

spevans committed:
SR-648: Allow swiftpm to statically link binaries on Linux

- If --build-swift-static-stdlib option is used then also produce
  static versions of libXCTest.a and libdispatch.a and put them
  into the lib/swift_static/linux/ toolchain directory.
  libFoundation.a is already being built and deployed there.

- Binaries with the swift libs statically linked in can then be
  built with using the command:

  swift build -Xswiftc -static-stdlib -Xlinker -lcurl -Xlinker -l:libxml2.a -Xlinker -llzma -Xlinker /lib/x86_64-linux-gnu/libz.so.1 -Xlinker -lbsd -Xlinker --allow-multiple-definition

  Note: This is a dynamic binary with the libswiftCore,
  libFoundation and libdispatch libraries statically linked in.

- Further fixes should reduce the complexity of the above command.

ddunbar commented:
@jrose-apple can you review this? We would like to get static-stdlib support into SwiftPM for 4 if possible.

The build-script changes LGTM, but the driver part I think needs you review (and @spevans it would be good to have a test encoding *why* this reordering is important, IMHO)

spevans commented:
The re-ordering of the auto link files is because when linking in static libraries and object files the ordering of arguments is important and the dependant libraries need to come after the objects/libs that require them. This is also the reason that the `-Xlinker` arguments are moved around.

My testing so far has not shown the changes to affect normal dynamic linking either. 

As for a test case I have a change for swift-corelibs-foundation that adds a static_test target that links the TestFoundation binary with the static swift libs which can be run manually on linux although I am still running some tests on it and it is not fully finished.

ddunbar commented:
I meant just a `swiftc` test case which validates the driver passes the Xlinker and auto linking arguments in the correct relative order.

spevans commented:
Oh I see, yes I should be able to add one for that.

jrose-apple commented:
I think if you're going to change it at all you should put the -Xlinker options all the way at the end. Anything else would run into some other ordering problem later.

spevans commented:
I moved the -Xlinker options as far to the end as was possible without causing the tests to break. This meant they ended up just before the `swift_end.o` used on ELF to mark the end of the `.swift` object sections.

I added a test to check the autolink and linker options are in the correct order.

jrose-apple commented:
Why do they need to be before swift_end.o? We don't expect `-Xlinker` options to include extra Swift object files.

(Also, why `--allow-multiple-definition`? And swiftc accepts `-l` flags without the `-Xlinker`.)

spevans commented:
I dont know the exact reason they need to be before the `swift_end.o` but when I moved them to just before the `-o` multiple tests failed. I did develop another solution where the linker args are added to a dummy object file to then be extracted as autolink arguments, which means that all of the extra linker commands can be hidden in one file. This would require some functionality to be added to swiftpm to dynamically generate the object file - currently I use the following script:

```
#!/bin/bash

output=$1
shift

read -r -d '' CODE <<EOF
// These commands will get extracted by 'swift-autolink-extract'
#define AUTOLINK_ENTRY(x) asm(".asciz \"" x "\"")

asm(".section .swift1_autolink_entries");
EOF

for arg in "$@"
do
    CODE="$CODE"$'\n'"AUTOLINK_ENTRY("\"${arg}\"");"
done
echo $CODE
clang++ -Wall -Werror -c -x c++ -o $output - <<<"$CODE"
```

Then it can be invoked as follows:
`swift-autolink-create static_args.o -Xlinker --allow-multiple-definition ...`

The `-Xlinker` args are actually going to `swift-build` not `swiftc` although it is possible to replace the `-Xlinker` with `-Xswiftc` except for when a direct path to a library is needed (which doesnt use `-l`).

The `--allow-multiple-definition` is just a temporary workaround because `objc_retainAutoreleasedReturnValue` is in both `libdispatch.a` and `libFoundation.a`. I do have a PR to have this function exported from `libdispatch.a`, then I can open another one to remove it from `libFoundation.a` and then this argument can be removed.

jrose-apple commented:
Ah, my bad on misreading the `-Xlinker`, and thanks for the explanation on `--allow-multiple-definition`. I'm still wondering what we're depending on in swiftc's handling of `-Xlinker`, though. Does swift-build pass `-Xlinker` arguments that need to appear early?

spevans committed:
[Linux] Reorder arguments for autolink files from swift-autolink-extract

- Reordering the auto link files when linking in static libraries
  and object files is required because the ordering of arguments
  is important, and the dependant libraries need to come after the
  objects/libs that require them. This is not a problem for
  libswiftCore.a but can be an issue with libs that sit on top of
  it, e.g. libFoundation.a

- Dont add an -rpath to the Swift dynamic libraries if using
  -static-stdlib

spevans committed:
[Linux] Reorder the -Xlinker and -l arguments

- Move the -Xlinker and -l options to just before swift_end.o

- Add testcase to check linker option ordering.

spevans commented:
The `-Xlinker --allow-multiple-definition` can now be dropped from the build command, it was fixed by the following PRs:
https://github.com/apple/swift-corelibs-libdispatch/pull/258
https://github.com/apple/swift-corelibs-foundation/pull/1090
https://github.com/apple/swift-corelibs-foundation/pull/1114

alblue commented:
@swift-ci please smoke test

spevans commented:
@alblue could you rerun the linux one please, from the logs it looks like it might have been a temporary network issue

ddunbar commented:
@swift-ci please smoke test linux

jrose-apple commented on test/Driver/linker-args-order-linux.swift:
```diff
@@ -0,0 +1,11 @@
+// REQUIRES: static_stdlib
+print("hello world!")
+// RUN: %empty-directory(%t)
+// RUN: %target-swiftc_driver -v -static-stdlib -o %t/static-stdlib %s -Xlinker --no-allow-multiple-definition 2>&1| %FileCheck %s
```
You shouldn't need to actually run the linker here. Instead, you can use `-###` (or `-driver-print-jobs`) to show what would be run. That way we can run the test on other platforms.

spevans replied:
Do you want me to update the test? The argument reordering change is only for Linux (actually GenericUnix) so I don't think it will make any difference on other platforms.

jrose-apple replied:
GenericUnix is still used in other places, and it's useful to be able to run the test even on Darwin. (Also, not actually compiling and linking would run faster.)

spevans replied:
@jrose-apple Ive added the -driver-print-jobs and removed the linux requirement. It tested ok on Linux and macOS. If the commit looks ok could you rerun the tests please.

jrose-apple reviewed (commented):
All right, I guess I'm okay with this. Thanks for sticking with it, Simon!

spevans committed:
Update the linker -Xlinker argument reordering test.

- Use -driver-print-jobs to avoid actually building the test file.

jrose-apple commented on test/Driver/linker-args-order-linux.swift:
```diff
@@ -1,11 +1,8 @@
-// RUN: %target-swiftc_driver -v -static-stdlib -o %t/static-stdlib %s -Xlinker --no-allow-multiple-definition 2>&1| %FileCheck %s
-// CHECK: Swift version
-// CHECK: Target: x86_64-unknown-linux-gnu
+// RUN: %target-swiftc_driver -driver-print-jobs -static-stdlib -o %t/static-stdlib %s -Xlinker --no-allow-multiple-definition 2>&1| %FileCheck %s
```
Thanks!

spevans commented:
Could someone please run a test on this latest change please

aciidb0mb3r commented:
@swift-ci please smoke test

aciidb0mb3r commented:
@swift-ci please smoke test

spevans commented:
@jrose-apple can this be merged now?

jrose-apple commented:
Ah, yes, looks good! Thanks, Simon.

jrose-apple commented:
If you want to bring this into Swift 4.0 branch, please follow the process at https://swift.org/blog/swift-4-0-release-process/. You can mention that this was reviewed by me.

ddunbar commented:
+1 on getting this into 4.

Thanks for your perseverance on this @spevans!
